src/pybind/mgr/rgw/module.py

```diff
@@ -0,0 +1,184 @@
+
+from . import *
+from .types import RGWAMException, RGWAMEnvMgr
+from .rgwam import EnvArgs, RGWAM
+
+
+class RGWAMOrchMgr(RGWAMEnvMgr):
+    def __init__(self, mgr):
+        self.mgr = mgr
+
+    def apply_rgw(self, svc_id, realm_name, zone_name, port = None):
+        spec = RGWSpec(service_id = svc_id,
+                       rgw_realm = realm_name,
+                       rgw_zone = zone_name,
+                       rgw_frontend_port = port)
+        completion = self.mgr.apply_rgw(spec)
+        orchestrator.raise_if_exception(completion)
+
+    def list_daemons(self, service_name, daemon_type = None, daemon_id = None, host = None, refresh = True):
+        completion = self.mgr.list_daemons(service_name,
+                                           daemon_type,
+                                           daemon_id=daemon_id,
+                                           host=host,
+                                           refresh=refresh)
+        return orchestrator.raise_if_exception(completion)
+
+
+class Module(orchestrator.OrchestratorClientMixin, MgrModule):
+    MODULE_OPTIONS = []
+
+    # These are "native" Ceph options that this module cares about.
+    NATIVE_OPTIONS = []
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        self.inited = False
+        self.lock = threading.Lock()
+        super(Module, self).__init__(*args, **kwargs)
+
+        # ensure config options members are initialized; see config_notify()
+        self.config_notify()
+
+        with self.lock:
+            self.inited = True
+            self.env = EnvArgs(RGWAMOrchMgr(self),
+                               str(self.get_ceph_conf_path()),
+                               f'mgr.{self.get_mgr_id()}',
+                               str(self.get_ceph_option('keyring')))
+
+        # set up some members to enable the serve() method and shutdown()
+        self.run = True
+        self.event = threading.Event()
+
+
+
+    def config_notify(self) -> None:
+        """
+        This method is called whenever one of our config options is changed.
+        """
+        # This is some boilerplate that stores MODULE_OPTIONS in a class
+        # member, so that, for instance, the 'emphatic' option is always
+        # available as 'self.emphatic'.
+        for opt in self.MODULE_OPTIONS:
+            setattr(self,
+                    opt['name'],
+                    self.get_module_option(opt['name']))
+            self.log.debug(' mgr option %s = %s',
+                           opt['name'], getattr(self, opt['name']))
+        # Do the same for the native options.
+        for opt in self.NATIVE_OPTIONS:
+            setattr(self,
+                    opt,
+                    self.get_ceph_option(opt))
+            self.log.debug(' native option %s = %s', opt, getattr(self, opt))
+
+
+    @CLICommand('rgw admin', perm='rw')
+    def _cmd_rgw_admin(self, params: Sequence[str]):
+        """rgw admin"""
+        run_cmd = [ './bin/radosgw-admin',
```

Comment:
where does `./bin/radosgw-admin` come from? does this expect a ceph build directory?

Reply:
When it runs as a mgr module, it's relative to the ceph build directory.

Reply:
isn't that going to be an issue for real deployments?

Reply:
should be running these commands with no directory (they're in $PATH) and with the args
```
            '-n', f"mgr.{self.mgr.get_mgr_id()}",
            '-k', str(self.mgr.get_ceph_option('keyring')),
```

Reply:
yeah, it passes these two plus the `-c` to point to the appropriate `ceph.conf`. I'll verify that it still works without the path (might have been needed for the cli tool).

Reply:
See: https://github.com/ceph/ceph/pull/42710/commits/3a65ad0cf74cf01b311ac3ec717faac28579b133